src/transport/PeerMessageCounter.h

```diff
@@ -135,6 +205,56 @@ class PeerMessageCounter
         }
     }
 
+    /**
+     * @brief
+     *    With the counter verified and the packet MIC also verified by the secure key, we can trust the packet and adjust
+     *    counter states including possible Rollover for Groups communications.
+     *
+     * @pre Verify(counter) == CHIP_NO_ERROR
+     */
+    void CommitWithRollOver(uint32_t counter)
+    {
+        if (counter <= mSynced.mMaxCounter)
+        {
+            // Take into account rollover
+            uint32_t offset         = mSynced.mMaxCounter - counter;
+            uint32_t rollOverOffset = ((UINT32_MAX - mSynced.mMaxCounter) + counter);
+            if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                mSynced.mWindow.set(offset);
+            }
+            else if (rollOverOffset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                // Roll over occured
+                mSynced.mMaxCounter = counter;
+                mSynced.mWindow <<= rollOverOffset;
+                mSynced.mWindow.set(0);
+            }
+            else
+            {
+                // Roll over occured with a leap bigger than CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE
+                mSynced.mMaxCounter = counter;
+                mSynced.mWindow.reset();
+                mSynced.mWindow.set(0);
+            }
+        }
+        else
+        {
+            uint32_t offset = counter - mSynced.mMaxCounter;
```

Comment:
So here we have `counter > mSynced.mMaxCounter`.  This may be because it's actually larger, or it may be because we're in a situation where `mSynced.mMaxCounter == 0`, say, due to rollover, and `counter` is `0xFFFFFFFF`, and that bit in the bitmap (that is, bit 0 or -1 depending on how we index into our bitmap) was not set.

And in this case, offset will be 0xFFFFFFFF, and the code below will clear the whole bitmap, which is _not_ the behavior we want.

Please add a test.

I think we want to do something similar to the verify case here in terms of using our mod 2^32 arithmetic instead of branching on `<` compares.  That is, compute:
```
    uint32_t pastOffset = mSynced.mMaxCounter - counter;
    if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE) {  // Or whatever the test should be based on whether we include the current counter in the window
        // Set the bit in the window
    } else {
        // Not a bit inside the window.  Since we are committing, this is a new mMaxCounter value.
        mSynced.mMaxCounter = counter;
        uint32_t shift = -pastOffset;
        // Shift the window by this much, or reset if shift is too big
    }
```
or something along those lines.

Reply:
Implemented and test added